### ISSUE TITLE
Hang when window is closed via button click

### ISSUE THREAD

**Issue by codingphil** [mozilla/geckodriver#941]
## System
* Version: 0.18
* Platform: Windows 10 Creators Update
* Firefox: 55.0.3 (64 bit)
* Selenium: 3.5.3

## Testcase

When calling click on a button which closes a window the click call hangs.

JUnit Selenium Test:
```java
  @Test
  public void testCloseWindowViaButtonClick() {
    // given
    driver.navigate().to("http://localhost/WindowOpen.html");
    Set<String> windowHandlesBeforeOpenWindow = driver.getWindowHandles();
    assertEquals(1, windowHandlesBeforeOpenWindow.size());
    String firstWindowHandle = windowHandlesBeforeOpenWindow.iterator().next();
    
    WebElement newWindowButton = driver.findElementById("btnNewWindow");
    newWindowButton.click();
    
    Set<String> windowHandlesAfterOpenWindow = driver.getWindowHandles();
    assertEquals(2, windowHandlesAfterOpenWindow.size());
    
    String secondWindowHandle = windowHandlesAfterOpenWindow.stream().filter(h -> !h.equals(firstWindowHandle)).findAny().get();
    driver.switchTo().window(secondWindowHandle);
    WebElement closeWindowButton = driver.findElementById("btnCloseWindow");
    
    // when
    closeWindowButton.click(); // <<==== HANGS
    
    // then
    Set<String> windowHandlesAfterCloseWindow = driver.getWindowHandles();
    assertEquals(1, windowHandlesBeforeOpenWindow.size());
    assertEquals(firstWindowHandle, windowHandlesAfterCloseWindow.iterator().next());
    driver.switchTo().window(firstWindowHandle);
    assertEquals("Window Open", driver.getTitle());
    assertEquals("Window Open", driver.findElementByTagName("h1").getText());
  }
```

**WindowOpen.html**
```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <title>Window Open</title>
        <meta charset="UTF-8">
    </head>
    <body>
        <h1>Window Open</h1>

        <button id="btnNewWindow" onClick="openNewWindow()">Open New Window</button>

        <script>
            function openNewWindow() {
                window.open("WindowClose.html", "_blank");
            }
        </script>
    </body>
</html>
```

**WindowClose.html:**
```html
<!DOCTYPE html>
<html>
    <head>
        <title>Window Close</title>
    </head>
    <body>
        <h1>Window Close</h1>
        <button id="btnCloseWindow" onClick="window.close()">Close Window</button>
    </body>
</html>
```

[SimpleWindowOpenClose.zip (HTML pages)](https://github.com/mozilla/geckodriver/files/1292145/SimpleWindowOpenClose.zip)

## Stacktrace

Stacktrace of hanging thread:

```java
Thread [Forwarding clickElement on session 67b03ca2-7d03-418a-8573-2f664a2baafa to remote] (Suspended)	
	SocketInputStream.socketRead0(FileDescriptor, byte[], int, int, int) line: not available [native method]	
	SocketInputStream.socketRead(FileDescriptor, byte[], int, int, int) line: 116	
	SocketInputStream.read(byte[], int, int, int) line: 170	
	SocketInputStream.read(byte[], int, int) line: 141	
	SessionInputBufferImpl.streamRead(byte[], int, int) line: 137	
	SessionInputBufferImpl.fillBuffer() line: 153	
	SessionInputBufferImpl.readLine(CharArrayBuffer) line: 282	
	DefaultHttpResponseParser.parseHead(SessionInputBuffer) line: 138	
	DefaultHttpResponseParser.parseHead(SessionInputBuffer) line: 56	
	DefaultHttpResponseParser(AbstractMessageParser<T>).parse() line: 259	
	LoggingManagedHttpClientConnection(DefaultBHttpClientConnection).receiveResponseHeader() line: 163	
	CPoolProxy.receiveResponseHeader() line: 165	
	HttpRequestExecutor.doReceiveResponse(HttpRequest, HttpClientConnection, HttpContext) line: 273	
	HttpRequestExecutor.execute(HttpRequest, HttpClientConnection, HttpContext) line: 125	
	MainClientExec.execute(HttpRoute, HttpRequestWrapper, HttpClientContext, HttpExecutionAware) line: 272	
	ProtocolExec.execute(HttpRoute, HttpRequestWrapper, HttpClientContext, HttpExecutionAware) line: 185	
	RetryExec.execute(HttpRoute, HttpRequestWrapper, HttpClientContext, HttpExecutionAware) line: 89	
	RedirectExec.execute(HttpRoute, HttpRequestWrapper, HttpClientContext, HttpExecutionAware) line: 111	
	InternalHttpClient.doExecute(HttpHost, HttpRequest, HttpContext) line: 185	
	InternalHttpClient(CloseableHttpClient).execute(HttpHost, HttpRequest, HttpContext) line: 72	
	InternalHttpClient(CloseableHttpClient).execute(HttpHost, HttpRequest, HttpContext) line: 56	
	ApacheHttpClient.fallBackExecute(HttpContext, HttpUriRequest) line: 139	
	ApacheHttpClient.execute(HttpRequest, boolean) line: 87	
	DriverCommandExecutor(HttpCommandExecutor).execute(Command) line: 161	
	DriverCommandExecutor.execute(Command) line: 82	
	FirefoxDriver(RemoteWebDriver).execute(String, Map<String,?>) line: 646	
	RemoteWebElement.execute(String, Map<String,?>) line: 275	
	RemoteWebElement.click() line: 82	
	WindowCloseViaButtonClickHangs.testCloseWindowViaButtonClick() line: 56	
```

## Trace-level log

[geckodriver-log.txt](https://github.com/mozilla/geckodriver/files/1292169/geckodriver-log.txt)

**Comment by andreastt** [comment]
It is not very helpful to report bugs on older Firefoxen that we cannot patch. Can you please attach a trace-level log using Firefox Nightly?

**Comment by codingphil** [comment]
The issue does not occur in Firefox Nightly and Firefox 56.0b10 any more. 
So we can close this issue.

What I noticed is that now the new windows opened via WebDriver are opened as tabs and not as windows (as before).

I then created another test page which opens a separate Firefox window via `window.open("WindowClose.html", "_blank", "height=300,width=300")` and the issue did not occur either.


[WindowOpenWithParameters-html.zip](https://github.com/mozilla/geckodriver/files/1294935/WindowOpenWithParameters-html.zip)
